<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [86]:
import pandas as pd
import psycopg2

In [87]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
from configparser import ConfigParser

# Считывание параметров подключения к базе данных из конфигурационного файла
config = ConfigParser()
config.read('config.ini')

host = config.get('postgresql', 'host')
port = config.get('postgresql', 'port')
database = config.get('postgresql', 'database')
user = config.get('postgresql', 'user')
password = config.get('postgresql', 'password')

In [88]:
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [89]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_3_1 = f'''select count(*)
                from vacancies
                '''
df = pd.read_sql_query(query_3_1, connection)
connection.close()
df.tail()

C:\Users\user\AppData\Local\Temp\ipykernel_18424\2879138304.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

In [90]:
# результат запроса
df.values
# array([[49197]], dtype=int64)

array([[49197]], dtype=int64)

2. Напишите запрос, который посчитает количество работодателей (таблица employers).

In [91]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_3_2 = f'''select distinct count(id)
                from employers
                '''
df = pd.read_sql_query(query_3_2, connection)
connection.close()
df.tail()

C:\Users\user\AppData\Local\Temp\ipykernel_18424\2300358882.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

In [92]:
# результат запроса
df.values
# array([[23501]], dtype=int64)

array([[23501]], dtype=int64)

3. Посчитайте с помощью запроса количество регионов (таблица areas).

In [93]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_3_3 = f'''select distinct count(id)
                from areas
                '''
df = pd.read_sql_query(query_3_3, connection)
connection.close()
df.tail()

C:\Users\user\AppData\Local\Temp\ipykernel_18424\1770019654.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

In [94]:
# результат запроса
df.values
# array([[1362]], dtype=int64)

array([[1362]], dtype=int64)

4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица industries).

In [95]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_3_4 = f'''select distinct count(id)
                from industries
                '''
df = pd.read_sql_query(query_3_4, connection)
connection.close()
df.tail()

C:\Users\user\AppData\Local\Temp\ipykernel_18424\1953924466.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

In [96]:
# результат запроса
df.values
# array([[294]], dtype=int64)

array([[294]], dtype=int64)

***

## Выводы по предварительному анализу данных:
Предварительный анализ данных, даёт нам ознакомительное представление о масштабах рынка труда, представленного в нашей базе, о большом количестве вакансий, распределенных по различным регионам и сферам деятельности. Количество работодателей явно превышает количество городов и сфер деятельности, что свидетельствует о высоком спросе и конкуренции на квалифицированных специалистов.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [97]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_4_1 = f'''select areas.name area,
	                count(areas.name) cnt
                from areas
                join vacancies v on v.area_id=areas.id
                group by area
                order by cnt desc
                '''
df = pd.read_sql_query(query_4_1, connection)
connection.close()
df.head(5)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\3951004122.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

In [98]:
# результат запроса
df.head(5).values

# array([['Москва', 5333],
#        ['Санкт-Петербург', 2851],
#        ['Минск', 2112],
#        ['Новосибирск', 2006],
#        ['Алматы', 1892]], dtype=object)

array([['Москва', 5333],
       ['Санкт-Петербург', 2851],
       ['Минск', 2112],
       ['Новосибирск', 2006],
       ['Алматы', 1892]], dtype=object)

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [99]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_4_2 = f'''select count(*)
                        from vacancies
                where salary_from is not null or salary_to is not null
                '''
df = pd.read_sql_query(query_4_2, connection)
connection.close()
df.head(5)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\2751696458.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

In [100]:
# результат запроса
df.values
# array([[24073]], dtype=int64)

array([[24073]], dtype=int64)

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [101]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_4_3 = f'''select
            	round(avg(salary_from)),
            	round(avg(salary_to))
            from vacancies
                '''
df = pd.read_sql_query(query_4_3, connection)
connection.close()
df.head(5)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\4033014717.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


round     round
0  71065.0  110537.0

In [102]:
# результат запроса
df.values
# array([[ 71065., 110537.]])

array([[ 71065., 110537.]])

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [103]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_4_4 = f'''select
            	count(name) cnt,
            	schedule,
            	employment
            from vacancies
            group by schedule, employment
            order by cnt desc
                '''
df = pd.read_sql_query(query_4_4, connection)
connection.close()
df.head(20)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\2331741631.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


cnt          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

In [104]:
# результат запроса
df.values
# array([[35367, 'Полный день', 'Полная занятость'],
#        [7802, 'Удаленная работа', 'Полная занятость'],
#        [1593, 'Гибкий график', 'Полная занятость'],
#        [1312, 'Удаленная работа', 'Частичная занятость'],
#        [940, 'Сменный график', 'Полная занятость'],
#        [569, 'Полный день', 'Стажировка'],
#        [367, 'Вахтовый метод', 'Полная занятость'],
#        [347, 'Полный день', 'Частичная занятость'],
#        [312, 'Гибкий график', 'Частичная занятость'],
#        [141, 'Полный день', 'Проектная работа'],
#        [133, 'Удаленная работа', 'Проектная работа'],
#        [116, 'Гибкий график', 'Стажировка'],
#        [101, 'Сменный график', 'Частичная занятость'],
#        [64, 'Удаленная работа', 'Стажировка'],
#        [18, 'Гибкий график', 'Проектная работа'],
#        [12, 'Сменный график', 'Стажировка'],
#        [2, 'Вахтовый метод', 'Проектная работа'],
#        [1, 'Сменный график', 'Проектная работа']], dtype=object)

array([[35367, 'Полный день', 'Полная занятость'],
       [7802, 'Удаленная работа', 'Полная занятость'],
       [1593, 'Гибкий график', 'Полная занятость'],
       [1312, 'Удаленная работа', 'Частичная занятость'],
       [940, 'Сменный график', 'Полная занятость'],
       [569, 'Полный день', 'Стажировка'],
       [367, 'Вахтовый метод', 'Полная занятость'],
       [347, 'Полный день', 'Частичная занятость'],
       [312, 'Гибкий график', 'Частичная занятость'],
       [141, 'Полный день', 'Проектная работа'],
       [133, 'Удаленная работа', 'Проектная работа'],
       [116, 'Гибкий график', 'Стажировка'],
       [101, 'Сменный график', 'Частичная занятость'],
       [64, 'Удаленная работа', 'Стажировка'],
       [18, 'Гибкий график', 'Проектная работа'],
       [12, 'Сменный график', 'Стажировка'],
       [2, 'Вахтовый метод', 'Проектная работа'],
       [1, 'Сменный график', 'Проектная работа']], dtype=object)

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [105]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_4_5 = f'''select
            	experience,
            	count(experience) cnt
            from vacancies
            group by experience
            order by cnt
                '''
df = pd.read_sql_query(query_4_5, connection)
connection.close()
df.head(20)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\2574478825.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

In [106]:
# результат запроса
df.values
# array([['Более 6 лет', 1337],
#        ['Нет опыта', 7197],
#        ['От 3 до 6 лет', 14511],
#        ['От 1 года до 3 лет', 26152]], dtype=object)

array([['Более 6 лет', 1337],
       ['Нет опыта', 7197],
       ['От 3 до 6 лет', 14511],
       ['От 1 года до 3 лет', 26152]], dtype=object)

***

### Выводы по детальному анализу вакансий
Выводы можно сделать следующие:
- Москва и Санкт-Петербург являются лидерами по количеству вакансий в России и занимают более 80% от общего числа. Минск и Алматы - самые крупные IT центры в Беларуси и Казахстане соответственно. Новосибирск - единственный город из Сибири в топ-5 по вакансиям.
- Примерно половина (49%) вакансий имеет заполненное хотя бы одно из двух полей с зарплатой. Это означает, что работодатели не всегда готовы раскрывать свой бюджет на определенную позицию или предлагать конкретную сумму до собеседования, вероятно это может говорить о том, что работодатели хотят точечно предлагать зарплату под каждого подходящего кандидата, в зависимости от его навыков, и диапозон может сильно отличаться.
- Работодатели которые указали зарплату прозрачны в вопросах компенсации и хотят привлечь кандидатов, которые ищут работу именно в этом диапазоне зарплат.
- Большинство работодателей предлагают полный день и полную занятость для IT специалистов. Это означает, что они ожидают высокого уровня ответственности и профессионализма от своих сотрудников. Удаленная работа и гибкий график стали более популярными в связи с пандемией коронавируса. Эти типы рабочего графика дают больше свободы и возможности для баланса между работой и личной жизнью.
- Самый редкий уровень опыта - более 6 лет. Это означает, что работодатели не часто ищут экспертов или лидеров в своих командах. Самый востребованный уровень опыта - от 1 года до 3 лет. Это означает, что работодатели предпочитают специалистов среднего уровня, которые уже имеют некоторый опыт работы и могут быстро адаптироваться к новым задачам и технологиям. Уровень опыта от 3 до 6 лет занимает второе место по количеству вакансий. Это означает, что работодатели также ценят специалистов высокого уровня, которые могут решать сложные проблемы и обладают глубокими знаниями в своей области. Уровень опыта - нет опыта, занимает третье место по количеству вакансий. Это означает, что работодатели также готовы давать шанс начинающим специалистам или студентам без предыдущего опыта работы.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [107]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_5_1 = f'''select
                	e.name,
                	count(e.name)
                from vacancies v
                join employers e on e.id=v.employer_id
                group by e.name
                order by 2 desc
                '''
df = pd.read_sql_query(query_5_1, connection)
connection.close()
df.head(20)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\1581397915.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0                     Яндекс   1933
1                 Ростелеком    491
2                   Тинькофф    444
3                       СБЕР    428
4              Газпром нефть    331
5                ИК СИБИНТЕК    327
6                        МТС    292
7                    DataArt    247
8      Совкомбанк Технологии    204
9                 Первый Бит    176
10                      Ozon    175
11                       IBS    175
12                    Тензор    155
13                  Andersen    155
14                Спецремонт    151
15                        VK    150
16    Поляков Денис Иванович    127
17  Иннотех, Группа компаний    122
18           Bell Integrator    121
19                     ANCOR    118

In [108]:
# результат запроса
df.iloc[:5].values
# array([['Яндекс', 1933],
#        ['Ростелеком', 491],
#        ['Тинькофф', 444],
#        ['СБЕР', 428],
#        ['Газпром нефть', 331]], dtype=object)

array([['Яндекс', 1933],
       ['Ростелеком', 491],
       ['Тинькофф', 444],
       ['СБЕР', 428],
       ['Газпром нефть', 331]], dtype=object)

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [109]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_5_2 = f'''select
                    a.name area,
                    count(distinct e.name) employers,
                    count(distinct v.name) vacancies
                from areas a
                left join employers e on a.id = e.area
                left join vacancies v on v.area_id = a.id
                where v.name is NULL
                group by a.name
                order by employers desc
                '''
df = pd.read_sql_query(query_5_2, connection)
connection.close()
df.head(10)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\782878935.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area  employers  vacancies
0                 Россия        410          0
1              Казахстан        207          0
2     Московская область         75          0
3     Краснодарский край         19          0
4               Беларусь         18          0
5     Ростовская область         18          0
6            Азербайджан         17          0
7   Республика Татарстан         16          0
8  Нижегородская область         16          0
9             Узбекистан         15          0

In [110]:
# результат запроса
df.iloc[:10].values
# array([['Россия', 410, 0],
#        ['Казахстан', 207, 0],
#        ['Московская область', 75, 0],
#        ['Краснодарский край', 19, 0],
#        ['Беларусь', 18, 0],
#        ['Ростовская область', 18, 0],
#        ['Азербайджан', 17, 0],
#        ['Республика Татарстан', 16, 0],
#        ['Нижегородская область', 16, 0],
#        ['Узбекистан', 15, 0]], dtype=object)

array([['Россия', 410, 0],
       ['Казахстан', 207, 0],
       ['Московская область', 75, 0],
       ['Краснодарский край', 19, 0],
       ['Беларусь', 18, 0],
       ['Ростовская область', 18, 0],
       ['Азербайджан', 17, 0],
       ['Республика Татарстан', 16, 0],
       ['Нижегородская область', 16, 0],
       ['Узбекистан', 15, 0]], dtype=object)

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [111]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_5_3 = f'''select
                    e.name Name_rabotodatel,
                    count(distinct v.area_id) Count_region
                from vacancies v
                left join employers e on e.id = v.employer_id
                group by e.name
                order by 2 desc
                '''
df = pd.read_sql_query(query_5_3, connection)
connection.close()
df.head(10)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\3131675368.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name_rabotodatel  count_region
0                      Яндекс           181
1                  Ростелеком           152
2                  Спецремонт           116
3      Поляков Денис Иванович            88
4                   ООО ЕФИН             71
5                  Совкомбанк            63
6                         МТС            55
7  ЭФКО, Управляющая компания            49
8                Почта России            48
9                        КРОН            48

In [112]:
# результат запроса
df.iloc[:10].values
# array([['Яндекс', 181],
#        ['Ростелеком', 152],
#        ['Спецремонт', 116],
#        ['Поляков Денис Иванович', 88],
#        ['ООО ЕФИН ', 71],
#        ['Совкомбанк', 63],
#        ['МТС', 55],
#        ['ЭФКО, Управляющая компания', 49],
#        ['Почта России', 48],
#        ['КРОН', 48]], dtype=object)

array([['Яндекс', 181],
       ['Ростелеком', 152],
       ['Спецремонт', 116],
       ['Поляков Денис Иванович', 88],
       ['ООО ЕФИН ', 71],
       ['Совкомбанк', 63],
       ['МТС', 55],
       ['ЭФКО, Управляющая компания', 49],
       ['Почта России', 48],
       ['КРОН', 48]], dtype=object)

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [113]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_5_4 = f'''select
                    count(*)
                from employers e
                left join EMPLOYERS_INDUSTRIES es on e.id = es.employer_id
                where es.industry_id is NULL
                '''
df = pd.read_sql_query(query_5_4, connection)
connection.close()
df.head(10)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\952322376.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

In [114]:
# результат запроса
df.values
# array([[8419]], dtype=int64)

array([[8419]], dtype=int64)

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [115]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_5_5 = f'''select
                	name,
                	count(industry_id)
                from employers e
                left join EMPLOYERS_INDUSTRIES es on e.id = es.employer_id
                group by name
                having count(es.industry_id) = 4
                order by name
                '''
df = pd.read_sql_query(query_5_5, connection)
connection.close()
df.head(3)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\1431302077.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  count
0  101 Интернет      4
1      21vek.by      4
2          2ГИС      4

In [116]:
# результат запроса
df.iloc[:3].values
# array([['101 Интернет', 4],
#        ['21vek.by', 4],
#        ['2ГИС', 4]], dtype=object)

array([['101 Интернет', 4],
       ['21vek.by', 4],
       ['2ГИС', 4]], dtype=object)

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [117]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_5_6 = f'''select count(*)
                from employers e
                left join employers_industries ei on ei.employer_id = e.id
                left join industries i on ei.industry_id = i.id
                where i.name = 'Разработка программного обеспечения'
                '''
df = pd.read_sql_query(query_5_6, connection)
connection.close()
df.head(3)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\2074747843.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

In [118]:
# результат запроса
df.values
# array([[3553]], dtype=int64)

array([[3553]], dtype=int64)

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [119]:
# код для получения списка городов-милионников

# Импортируем библиотеку для работы с веб-страницами
import requests
# Импортируем библиотеку для парсинга HTML
from bs4 import BeautifulSoup

# Задаем ссылку на веб-страницу
url = "https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8"

# Получаем содержимое веб-страницы
response = requests.get(url)

# Проверяем статус ответа
if response.status_code == 200:
    # Преобразуем содержимое веб-страницы в объект BeautifulSoup
    soup = BeautifulSoup(response.text, "html.parser")
    # Находим таблицу с данными о городах-милионниках используя индекс таблицы
    table = soup.find_all('table')[1]
    # Находим все строки таблицы
    rows = table.find_all("tr")
    # Создаем пустой список для хранения названий городов
    cities = []
    # Проходим по всем строкам таблицы, начиная со второй (первая - заголовок)
    for row in rows[1:]:
        # Находим все ячейки строки
        cells = row.find_all("td")
        # Получаем название города из первой ячейки строки и удаляем пробелы и переносы строк по краям
        city = cells[1].text.strip()
        # Добавляем название города в список
        cities.append(city)
else:
    # Выводим сообщение об ошибке, если статус ответа не равен 200 (OK)
    print(f"Ошибка при получении данных: {response.status_code}")

# Проходим по всем элементам списка
for i in range(len(cities)):
    # Если элемент содержит "[a]"
    if "[a]" in cities[i]:
        # Заменяем "[a]" на пустую строку
        cities[i] = cities[i].replace("[a]", "")
# Преобразуем список в кортеж
cities = tuple(cities)
# Выводим измененный список на экран
print(cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')


In [120]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_5_7 = f'''WITH temp AS(
                        select
                        	a.name,
                        	count(*) cnt
                        from vacancies v
                        left join areas a on v.area_id=a.id
                        left join employers e on v.employer_id=e.id
                        where e.name='Яндекс' and a.name in {cities}
                        group by a.name
                        order by 2)
                SELECT *  FROM temp
                UNION ALL
                SELECT 'Total', sum(cnt)  FROM temp
                '''
df = pd.read_sql_query(query_5_7, connection)
connection.close()
df.head(20)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\329275006.py:24: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7               Уфа   26.0
8            Самара   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

In [121]:
# результат запроса
df.iloc[:20].values
# array([['Омск', 21.0],
#        ['Челябинск', 22.0],
#        ['Красноярск', 23.0],
#        ['Волгоград', 24.0],
#        ['Пермь', 25.0],
#        ['Казань', 25.0],
#        ['Ростов-на-Дону', 25.0],
#        ['Уфа', 26.0],
#        ['Самара', 26.0],
#        ['Краснодар', 30.0],
#        ['Воронеж', 32.0],
#        ['Новосибирск', 35.0],
#        ['Нижний Новгород', 36.0],
#        ['Екатеринбург', 39.0],
#        ['Санкт-Петербург', 42.0],
#        ['Москва', 54.0],
#        ['Total', 485.0]], dtype=object)

array([['Омск', 21.0],
       ['Челябинск', 22.0],
       ['Красноярск', 23.0],
       ['Волгоград', 24.0],
       ['Пермь', 25.0],
       ['Казань', 25.0],
       ['Ростов-на-Дону', 25.0],
       ['Уфа', 26.0],
       ['Самара', 26.0],
       ['Краснодар', 30.0],
       ['Воронеж', 32.0],
       ['Новосибирск', 35.0],
       ['Нижний Новгород', 36.0],
       ['Екатеринбург', 39.0],
       ['Санкт-Петербург', 42.0],
       ['Москва', 54.0],
       ['Total', 485.0]], dtype=object)

***

### Выводы по анализу работодателей
Выводы можно сделать следующие:
- Самое большое количество вакансий мы видем у Яндекса, это говорит о том что Яндекс уже достаточно давно на рынке и у них очень много продуктов и проектов, поэтому им требуется так много специалистов.
- Рынок труда в регионе, с большим количеством работодателей, но небольшим количеством вакансий, может быть особенно сложным. Это может быть связано с различными факторами, такими как отсутствие экономического роста, низкий спрос на определенные виды работ или высокая конкуренция на рынке труда. Работодатели в этом регионе, возможно, больше сосредоточены на сохранении имеющейся рабочей силы, чем на расширении и найме новых сотрудников. Это может быть вызвано различными причинами, такими как напряженный рынок труда, высокая текучесть кадров или ориентация на эффективность, а не на рост. Также в регионе может быть большое количество малых предприятий или стартапов, которые еще не достигли той стадии, когда они нанимают новых сотрудников.
- Опять же мы видим что Яндекс находится в лидерах по количеству регионов публикуемых вакансий и является самым активным работодателем, что может свидетельствовать об их широкой географической деятельности и потребности в различных специалистах.
- 8419 - столь большое количество работодателей, у которых не указана сфера деятельности, может быть связано с разными причинами, например: Работодатели не хотят раскрывать свою конфиденциальную информацию или конкурентные преимущества; Работодатели не знают, как правильно классифицировать свою сферу деятельности или не находят подходящую категорию в списке.
- Количество компаний, у которых указано четыре сферы деятельности - это 1136 из общего количества 23501. Это составляет около 4.8% от всех работодателей. Это может говорить о том, что эти компании имеют более широкий спектр деятельности или более сложную структуру организации по сравнению с другими работодателями.
- Количество работодателей, у которых в качестве сферы деятельности указана «Разработка программного обеспечения» - это 3553 из общего количества 23501. Это составляет около 15.1% от всех работодателей. Это может говорить о том, что разработка программного обеспечения является одной из самых популярных и востребованных сфер деятельности на рынке труда. Разработка программного обеспечения может быть связана с разными отраслями и направлениями, такими как информационные технологии, интернет-сервисы, игровая индустрия, финансовые технологии и другие. Это может говорить о том, что работодатели из этих сфер нуждаются в квалифицированных специалистах по разработке программного обеспечения для решения своих задач и целей.
- По пунку 7, можно сделать следующие выводы:
    - Яндекс предлагает вакансии во всех городах-миллионниках России.
    - Больше всего вакансий Яндекса в Москве и Санкт-Петербурге, что соответствует статусу этих городов как крупных экономических и научных центров.
    - Меньше всего вакансий Яндекса в Омске и Челябинске, что может быть связано с меньшим спросом на услуги Яндекса или конкуренцией с другими компаниями в этих регионах.
    - Общее количество вакансий Яндекса составляет 485, что свидетельствует о том, что компания активно развивается и нуждается в новых сотрудниках.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [122]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_6_1 = f'''select count(*)
                from vacancies
                where name ilike '%data%' or name ilike '%данн%'
                '''
df = pd.read_sql_query(query_6_1, connection)
connection.close()
df.head(20)

C:\Users\user\AppData\Local\Temp\ipykernel_18424\3305081415.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

In [123]:
# результат запроса
df.values
# array([[1771]], dtype=int64)

array([[1771]], dtype=int64)

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [124]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_6_2 = f'''select count(*)
                from vacancies
                where (name ilike '%data scientist%'
                	or name ilike '%data science%'
                	or name ilike '%исследователь данных%'
                	or (name ilike '%ML%' and name not ilike '%HTML%')
                	or name ilike '%machine learning%'
                	or name ilike '%машинн%обучен%')
                	and (name ilike '%junior%'
                	or experience ilike '%Нет опыта%'
                	or employment ilike '%Стажировка%')
                '''
df = pd.read_sql_query(query_6_2, connection)
connection.close()
df.head()

C:\Users\user\AppData\Local\Temp\ipykernel_18424\2868881126.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

In [125]:
# результат запроса
df.values
# array([[51]], dtype=int64)

array([[51]], dtype=int64)

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [126]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_6_3 = f'''select count(*)
                from vacancies
                where (name ilike '%data scientist%'
                	or name ilike '%data science%'
                	or name ilike '%исследователь данных%'
                	or (name like '%ML%' and name not ilike '%HTML%')
                	or name ilike '%machine learning%'
                	or name ilike '%машинн%обучен%')
                and (key_skills ilike '%sql%' or key_skills ilike '%postgres%')
                '''
df = pd.read_sql_query(query_6_3, connection)
connection.close()
df.head()

C:\Users\user\AppData\Local\Temp\ipykernel_18424\3584441375.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

In [127]:
# результат запроса
df.values
# array([[201]], dtype=int64)

array([[201]], dtype=int64)

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [128]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_6_4 = f'''select count(*)
                from vacancies v
                left join areas a on v.area_id=a.id
                left join employers e on v.employer_id=e.id
                where (v.name ilike '%data scientist%'
                	or v.name ilike '%data science%'
                	or v.name ilike '%исследователь данных%'
                	or (v.name like '%ML%' and v.name not ilike '%HTML%')
                	or v.name ilike '%machine learning%'
                	or v.name ilike '%машинн%обучен%')
                and key_skills ilike '%python%'
                '''
df = pd.read_sql_query(query_6_4, connection)
connection.close()
df.head()

C:\Users\user\AppData\Local\Temp\ipykernel_18424\1905390275.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

In [129]:
# результат запроса
df.values
# array([[351]], dtype=int64)

array([[351]], dtype=int64)

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [130]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_6_5 = f'''select
                	round(avg(LENGTH(key_skills)-LENGTH(REPLACE(key_skills,CHR(9),''))+1),2)
                from vacancies v
                left join areas a on v.area_id=a.id
                left join employers e on v.employer_id=e.id
                where (v.name ilike '%data scientist%'
                	or v.name ilike '%data science%'
                	or v.name ilike '%исследователь данных%'
                	or (v.name like '%ML%' and v.name not ilike '%HTML%')
                	or v.name ilike '%machine learning%'
                	or v.name ilike '%машинн%обучен%')
                and key_skills IS NOT NULL
                '''
df = pd.read_sql_query(query_6_5, connection)
connection.close()
df.head()

C:\Users\user\AppData\Local\Temp\ipykernel_18424\1476652805.py:23: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

In [131]:
# результат запроса
df.values
# array([[6.41]])

array([[6.41]])

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [132]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_6_6 = f'''SELECT
                  ROUND(AVG(COALESCE((salary_from+salary_to)/2, salary_from, salary_to))) AS Average_Salary,
                  experience
                FROM vacancies v
                LEFT JOIN areas a ON v.area_id = a.id
                LEFT JOIN employers e ON v.employer_id = e.id
                WHERE (v.name ILIKE '%data scientist%'
                  OR v.name ILIKE '%data science%'
                  OR v.name ILIKE '%исследователь данных%'
                  OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
                  OR v.name ILIKE '%machine learning%'
                  OR v.name ILIKE '%машинн%обучен%')
                  AND (salary_from IS NOT NULL
                  OR salary_to IS NOT NULL)
                GROUP BY experience
            '''
df = pd.read_sql_query(query_6_6, connection)
connection.close()
df.head()

C:\Users\user\AppData\Local\Temp\ipykernel_18424\309804637.py:26: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


average_salary          experience
0         74643.0           Нет опыта
1        139675.0  От 1 года до 3 лет
2        243115.0       От 3 до 6 лет

In [133]:
# результат запроса
df.values
# array([[74643.0, 'Нет опыта'],
#        [139675.0, 'От 1 года до 3 лет'],
#        [243115.0, 'От 3 до 6 лет']], dtype=object)

array([[74643.0, 'Нет опыта'],
       [139675.0, 'От 1 года до 3 лет'],
       [243115.0, 'От 3 до 6 лет']], dtype=object)

### Дополнительно найду вакансии которые в Питере (т.к. я сам тут живу) или с возможностью удалённой работы из любого города
Только вакансии у которы опыт не более 3-х лет

In [135]:
# текст запроса
connection = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

query_6_6 = f'''select
                	*
                from vacancies v
                left join areas a on v.area_id=a.id
                left join employers e on v.employer_id=e.id
                where (v.name ilike '%data scientist%'
                	or v.name ilike '%data science%'
                	or v.name ilike '%исследователь данных%'
                	or (v.name like '%ML%' and v.name not ilike '%HTML%')
                	or v.name ilike '%machine learning%'
                	or v.name ilike '%машинн%обучен%')
                and key_skills IS NOT NULL
                and key_skills ilike '%python%'
                and (a.name ilike '%санкт%' or v.schedule ilike '%удал%')
                and v.experience not ilike '%От 3%'
            '''
df = pd.read_sql_query(query_6_6, connection)
connection.close()
df.head()

C:\Users\user\AppData\Local\Temp\ipykernel_18424\3056556698.py:26: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


id                                            name  \
0  55351195                                  Data Scientist   
1  55166994                       ML разработчик (удалённо)   
2  55529384  Математик, Junior/Middle/Senior Data Scientist   
3  55145641                           Junior Data Scientist   
4  44095047  Инженер-тестировщик в команду Machine Learning   

                                          key_skills          schedule  \
0   Python\tSQL\tNLP\tScikit-learn\tMachine Learning       Полный день   
1  Python\tПоиск информации в интернет\tC++\tМате...  Удаленная работа   
2  Python\tMATLAB\tМатематическое моделирование\t...  Удаленная работа   
3  Python\tSQL\tMachine Learning\tData Science\tА...       Полный день   
4         Python\tGit\tAtlassian Jira\tLinux\tMATLAB  Удаленная работа   

           experience           employment  salary_from  salary_to  area_id  \
0  От 1 года до 3 лет     Полная занятость          NaN        NaN        2   
1  От 1 года до 3 лет     Полная занятость      50000.0   100000.0        1   
2  От 1 года до 3 лет  Частичная занятость          NaN        NaN     1002   
3           Нет опыта     Полная занятость      50000.0        NaN        2   
4  От 1 года до 3 лет     Полная занятость          NaN        NaN        4   

   employer_id    id             name       id  \
0         3127     2  Санкт-Петербург     3127   
1      6154479     1           Москва  6154479   
2      1688252  1002            Минск  1688252   
3        40938     2  Санкт-Петербург    40938   
4         8550     4      Новосибирск     8550   

                                                name  area  
0                                            МегаФон     1  
1                                  Cpaaggregator.com     1  
2  Представительство ООО «Квалитет» (Российская Ф...  1002  
3                                      Улыбка радуги     2  
4                        Центр финансовых технологий     4

In [142]:
# результат запроса
df['id'].count()

id    73
id    73
id    73
dtype: int64

***

### Выводы по предметному анализу
Выводы можно сделать следующие:
- Запрос показывает, что вакансии, связанные с данными, составляют около 3.6% от общего числа вакансий в России. Это может означать, что спрос на специалистов по данным невысок или что такие специалисты редки и ценны.
- Запрос показывает, что вакансии для начинающих дата-сайентистов составляют около 2.9% от общего числа вакансий, связанных с данными. Это может означать, что конкуренция за такие вакансии довольно высока и требуются хорошие навыки и опыт.
- Запрос показывает, что вакансии для data scientist, в которых требуются навыки SQL или postgres, составляют около 11.4% от общего числа вакансий, связанных с данными. Это может означать, что такие навыки довольно востребованы и ценятся работодателями.
- Запрос показывает, что вакансии для data scientist, в которых требуется знание Python, составляют около 19.8% от общего числа вакансий, связанных с данными. Это может означать, что Python является одним из самых популярных и востребованных языков программирования для работы с данными.
- Запрос показывает, что в среднем работодатели указывают 6.41 ключевых навыков в вакансиях для data scientist. Это может означать, что работа data scientist требует широкого спектра знаний и умений в разных областях.
- Запрос показывает, что работа data scientist является высокооплачиваемой и ценится на рынке труда. Также это может означать, что чем больше у вас опыта и навыков, тем выше ваша потенциальная зарплата.

# Общий вывод по проекту

- Рынок IT вакансий довольно разнообразен и предлагает разные условия работы, зарплаты и требования к опыту и навыкам. Большинство работодателей предпочитают специалистов среднего и высокого уровня опыта, которые могут работать полный день и полную занятость. Однако также есть возможности для начинающих специалистов или тех, кто хочет работать удаленно или по гибкому графику. Зарплата по вакансиям зависит от многих факторов, таких как регион, город, сфера деятельности, тип рабочего графика и трудоустройства и требуемый уровень опыта. В целом, чем выше эти параметры, тем выше зарплата.

- Проект показал, что спрос на data scientist в России высок и зависит от региона, опыта и навыков кандидатов. SQL или postgres являются одними из самых востребованных и ценных навыков для дата-сайентистов. Это говорит о том, что работодатели ожидают от дата-сайентистов умения работать с базами данных и извлекать из них полезную информацию для анализа и прогнозирования. Средняя зарплата для data scientist растет с увеличением опыта работы и достигает более 200 тысяч рублей в месяц для опытных специалистов. Работодатели предпочитают полный рабочий день и полную занятость, но также есть возможность работать удаленно или частично.

- Разработка программного обеспечения является одной из самых популярных и востребованных сфер деятельности на рынке IT вакансий. Это говорит о том, что работодатели из разных отраслей и направлений нуждаются в квалифицированных специалистах по разработке программного обеспечения для решения своих задач и целей. Разработка программного обеспечения может быть связана с разными видами продуктов и проектов, такими как информационные технологии, интернет-сервисы, игровая индустрия, финансовые технологии и другие.

- Яндекс является одним из самых крупных и активных работодателей на рынке IT вакансий. Яндекс предлагает вакансии во всех городах-миллионниках России. Больше всего вакансий Яндекса в Москве и Санкт-Петербурге, что соответствует статусу этих городов как крупных экономических и научных центров. Меньше всего вакансий Яндекса в Омске и Челябинске, что может быть связано с меньшим спросом на услуги Яндекса или конкуренцией с другими компаниями в этих регионах. Общее количество вакансий Яндекса составляет 485, что свидетельствует о том, что компания активно развивается и нуждается в новых сотрудниках.

- Python является одним из самых популярных и востребованных языков программирования для работы с данными. Это говорит о том что Python предлагает много возможностей для дата-сайентистов в плане анализа, визуализации и машинного обучения. С помощью Python можно работать с разными форматами и источниками данных, применять различные статистические и математические методы, создавать интерактивные графики и диаграммы, обучать и тестировать модели машинного обучения и нейронных сетей. Python также поддерживает множество библиотек и фреймворков для работы с данными, таких как pandas, numpy, scipy, matplotlib, seaborn, scikit-learn, tensorflow и другие.